Decouple TimeFormatLocalization cluster

tersal opened the pull request:
#### Summary

Proposal for code driven change for TimerFormatLocalization cluster.

This change separates the code in Cluster and Logic layers, since it is a small cluster I expect not a lot of additional memory usage (or at least near the expected additional usage).

Two attributes use persistence storage (ActiveCalendarType and HourFormat)

#### Related issues

Fixes #39797 
Depends-On: #40568

#### Testing

Tested writing and reading values manually using chip-repl, the attributes seem to be stored, retrieved and displayed properly.

Tested writing/reading to persistence using `AttributePersistence` and `WriteValue` directly from `AttributeProvider` the combination of both seem to work properly. Currently using `WriteValue` for `ActiveCalendarType` writes to avoid storing an unsupported value after decoding.

tersal committed:
Add initial TimeFormatLocalization decoupling

tersal committed:
Add comments and changes to build test file

tersal committed:
Add code improvements from internal feedback

tersal committed:
Add FeatureFlags handling

tersal committed:
Add changes to UnitTests

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Add test persistence implementation

tersal committed:
minor code fixes

tersal committed:
revert local testing changes un python build

tersal commented on scripts/build_python.sh:
```diff
@@ -229,9 +229,9 @@ gn_args=(
+    "chip_crypto=\"boringssl\""
     "chip_build_controller_dynamic_server=$chip_build_controller_dynamic_server"
-    "chip_support_webrtc_python_bindings=true"
+    "chip_support_webrtc_python_bindings=false"
```
Changes unrelated to this specific PR and just used for local testing.

Removed.

tersal committed:
Move WriteAttribute to Logic layer

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Start Read/Write UnitTest implementation using AttributePersistenceProvider

tersal committed:
Update tests cases

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

tersal committed:
reorganize code and add basic validations

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,128 @@
+    // If initial value is not one of the allowed values, pick one valid value and write it.
+    TimeFormatLocalization::CalendarTypeEnum validCalendar = TimeFormatLocalization::CalendarTypeEnum::kBuddhist;
+    TimeFormatLocalizationLogic::IsSupportedCalendarType(calendarType, &validCalendar);
+    mLogic.setActiveCalendarType(validCalendar);
```
Added validation for ActiveCalendarType initial read.

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,128 @@
+    MutableByteSpan hourBytes(reinterpret_cast<uint8_t *>(&hourFormat), sizeof(hourFormat));
+    error = context.attributeStorage->ReadValue({kRootEndpointId, TimeFormatLocalization::Id, TimeFormatLocalization::Attributes::HourFormat::Id},
+                                                            hourBytes);
+    mLogic.setHourFormat(hourFormat);
```
Added validation for HourFormat attribute Read operation.

tersal commented on src/app/clusters/time-format-localization-server/tests/SampleDeviceProvider.cpp:
```diff
@@ -0,0 +1,139 @@
+    return kNumSupportedLocales;
+}
+
+bool SampleDeviceProvider::SupportedLocalesIteratorImpl::Next(CharSpan & output)
+{
+    return true;
+}
```
SupportedLocales are not implemented and not used for the code of these unit tests. Just here to ensure compilation.

tersal commented on src/app/clusters/time-format-localization-server/tests/SampleDeviceProvider.cpp:
```diff
@@ -0,0 +1,139 @@
+
+}
+
+bool SampleDeviceProvider::UserLabelIteratorImpl::Next(UserLabelType & output)
+{
+    return true;
+}
```
These Provider is only used for the TimeFormatLocalization cluster Unit Tests, LabelIterator is not used in that cluster thus not needed in the "test"provider.

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Change initialization list in constructor

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

tersal committed:
Reduce SampleProvider complexity and size for UnitTests

tersal committed:
Add unit test for invalid values

tersal committed:
Minor changes for constant elements

tersal commented on src/app/clusters/time-format-localization-server/tests/SampleDeviceProvider.cpp:
```diff
@@ -0,0 +1,135 @@
+bool SampleDeviceProvider::UserLabelIteratorImpl::Next(UserLabelType & output)
+{
+    return true;
+}
```
This specific implementation won't be used outside UnitTests and only for CalendarTypes, so these are unused stubs needed to compile the Provider.

tersal commented on src/app/clusters/time-format-localization-server/tests/SampleDeviceProvider.cpp:
```diff
@@ -0,0 +1,135 @@
+bool SampleDeviceProvider::SupportedLocalesIteratorImpl::Next(CharSpan & output)
+{
+    return true;
+}
```
This specific implementation won't be used outside UnitTests and only for CalendarTypes, so these are unused stubs needed to compile the Provider.

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,250 @@
+CHIP_ERROR TimeFormatLocalizationLogic::Attributes(ReadOnlyBufferBuilder<DataModel::AttributeEntry> & builder) const
+{
+    // Ensure capacity just in case
+    ReturnErrorOnFailure(
```
use https://github.com/project-chip/connectedhomeip/blob/master/src/app/server-cluster/AttributeListBuilder.h#L36

Generates smaller code and does not need to pre-allocate storage (does so automatically)

tersal committed:
Add changes for AttributeListBuilder

tersal committed:
Changes to AttributePersistenceProvider in Logic

andy31415 commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,87 @@
+    gServer.Destroy();
+}
+
+Protocols::InteractionModel::Status MatterTimeFormatLocalizationClusterServerPreAttributeChangedCallback(
```
This should be deleted probably. Remove the cluster from https://github.com/project-chip/connectedhomeip/blob/master/src/app/common/templates/config-data.yaml#L113 and delete this method.

tersal replied:
Removed callback from the config-data file and regenerated files, also removed InitFunction `emberAfTimeFormatLocalizationClusterServerInitCallback` from the same file.

andy31415 commented on src/app/clusters/time-format-localization-server/tests/TestTimeFormatLocalizationCluster.cpp:
```diff
@@ -0,0 +1,347 @@
+
+    {
+        // No attributes enabled
+        BitFlags<TimeFormatLocalization::Feature> features{ 0 };
+        features.Set(TimeFormatLocalization::Feature::kCalendarFormat);
```
Can we do this on one go?

maybe

```suggestion
        const BitFlags<TimeFormatLocalization::Feature> features(TimeFormatLocalization::Feature::kCalendarFormat);
```

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,104 @@
+ *    limitations under the License.
+ */
+
+#include "time-format-localization-cluster.h"
```
I think we should use the full `<>` path as a standard if possible.

tersal replied:
Fixed.

tersal committed:
Fix failures in unit test due change in Setters

tersal committed:
Removed ember related callbacks

tersal commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,87 @@
+
+}
+
+void emberAfTimeFormatLocalizationClusterServerInitCallback(EndpointId endpoint) {}
```
One of them seems to be generated based on the data in [config-data.yaml](https://github.com/project-chip/connectedhomeip/blob/master/src/app/common/templates/config-data.yaml#L84). I removed that one too and deleted this function.

tersal committed:
Change BitFlags initialization in Tests

tersal committed:
Change include header format

tersal committed:
Remove curly braces from switch statements when applicable

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,104 @@
+        }
+        return result;
+    }
+    default: {
```
Removed.

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,104 @@
+{
+    switch (request.path.mAttributeId)
+    {
+    case TimeFormatLocalization::Attributes::HourFormat::Id: {
```
Removed.

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
test integration of AttributePersistence

tersal committed:
Combine change individual setter to a combined attribute handler in Logic

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,245 @@
+    VerifyOrReturn(attrProvider != nullptr);
+    CalendarTypeEnum calendarType = kDefaultCalendarType;
+
+    // Try to read existing calendar type from persistence
```
Updated to use `LoadNativeEndianValue` and `DecodeAndStoreNativeEndianValue` from `AttributePersistence`

tersal committed:
Removed test logs, and simplified code

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,104 @@
+        return mLogic.GetSupportedCalendarTypes(encoder);
+    }
+    case TimeFormatLocalization::Attributes::FeatureMap::Id: {
+        return encoder.Encode(mLogic.GetFeatureMap().Raw());
```
Removed raw().

tersal committed:
Change zap generation for ClusterRevision

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by prettier-json

tersal commented on examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:
```diff
@@ -7582,8 +7582,8 @@ endpoint 0 {
-    persist  attribute hourFormat default = 0;
-    persist  attribute activeCalendarType default = 0;
+    callback attribute hourFormat default = 0;
+    callback attribute activeCalendarType default = 0;
```
Changed the cluster revision in the zcl files.

tersal committed:
Remove breaking UnitTest

tersal commented:
/gemini review

bzbarsky-apple commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,78 @@
+
+}
+
+void emberAfTimeFormatLocalizationClusterInitCallback(EndpointId endpoint)
```
Please see https://github.com/project-chip/connectedhomeip/pull/39698#discussion_r2264527455 -- this is wrong.  Same for the shutdown callback.

tersal committed:
Change ember callbacks

tersal committed:
Change calendar setter to avoid double decoding

tersal committed:
Fix ActiveCalendarType not being stored properly

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Remove test changes

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

tersal committed:
Change type of AttributePersistenProvider in Logic layer

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,223 @@
+            { kRootEndpointId, TimeFormatLocalization::Id, ActiveCalendarType::Id }, decoder, mCalendarType);
+
+        return result == CHIP_NO_ERROR ? Protocols::InteractionModel::Status::Success
+                                       : Protocols::InteractionModel::Status::WriteIgnored;
```
Changed to use WriteAttribute directly to avoid double decoding and ensure that an unsupported value is not stored.

andy31415 commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,78 @@
+using namespace chip::app::Clusters;
+using namespace Protocols::InteractionModel;
+
+using namespace chip::app::Clusters::TimeFormatLocalization::Attributes;
```
do we need this using?

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.h:
```diff
@@ -0,0 +1,74 @@
+{
+public:
+    static constexpr TimeFormatLocalization::CalendarTypeEnum kDefaultCalendarType =
+        TimeFormatLocalization::CalendarTypeEnum::kBuddhist;
```
spec does not seem to have a format and this seems like an odd default to have (maybe? no idea what the correct one if any is).

I expect previous applications to have used zap to select the right version. How about we keep the zap data and we query it in codegen integration to set it in the constructor?

It is unfortunate to need to pay the RAM cost for startup initialization, however this seems to be what applications would expect. In that case calnedar type and hour format would stay as RAM in zap.

tersal replied:
Added logic to read from RAM the values for HourFormat and CalendarType at startup.

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.h:
```diff
@@ -0,0 +1,74 @@
+    // Attributes handling for cluster
+    TimeFormatLocalization::HourFormatEnum GetHourFormat() const;
+    TimeFormatLocalization::CalendarTypeEnum GetActiveCalendarType() const;
+    DataModel::ActionReturnStatus WriteAttribute(const DataModel::WriteAttributeRequest & request,
```
if we do this, we may as well just implement the cluster. The intent is either:
- cluster and that is it OR
- logic is always type safe, cluster converts into type safe. So `AttributeValueDecoder` is not typesafe and should reside in cluster.

tersal replied:
It probably makes more sense to move everything to the cluster, this seems to be a rather small cluster.

andy31415 commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,78 @@
+        rawFeatureMap = 0;
+    }
+
+    gServer.Create(endpoint, BitFlags<TimeFormatLocalization::Feature>(rawFeatureMap));
```
I believe we should load the defaults for calendar type and hour format from zap as well, since it seems reasonable for that to be product specific. 

In that case we should keep those items in RAM and allow accessors to be available.

tersal replied:
Changed.

andy31415 commented on src/app/clusters/time-format-localization-server/app_config_dependent_sources.gni:
```diff
@@ -11,4 +11,4 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-app_config_dependent_sources = [ "time-format-localization-server.cpp" ]
```
PR also needs to delete the old files

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    case TimeFormatLocalization::Attributes::ActiveCalendarType::Id: {
+        if (!mFeatures.Has(TimeFormatLocalization::Feature::kCalendarFormat))
```
API requirement is that writes are done only on items return by ::Attributes. This check is not needed.

tersal replied:
Removing check.

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,222 @@
+        CHIP_ERROR result = attrProvider.WriteValue({ kRootEndpointId, TimeFormatLocalization::Id, ActiveCalendarType::Id },
+                                                    { reinterpret_cast<const uint8_t *>(&mCalendarType), sizeof(mCalendarType) });
+
+        return result == CHIP_NO_ERROR ? Protocols::InteractionModel::Status::Success
```
same here.

andy31415 commented on src/app/persistence/AttributePersistence.h:
```diff
@@ -43,7 +43,7 @@ class AttributePersistence
     /// Error reason for load failure is logged (or nothing logged in case "Value not found" is the
     /// reason for the load failure).
-    template <typename T, typename std::enable_if_t<std::is_arithmetic_v<T>> * = nullptr>
+    template <typename T, typename std::enable_if_t<std::is_arithmetic_v<T> || std::is_enum_v<T>> * = nullptr>
```
we can extract this into a separate PR.

Could we have unit tests on it as well?

tersal replied:
Moved to a different PR #40568 and added unit test.

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Regenerate ZAP code

tersal committed:
Move logic to cluster

tersal committed:
Remove unused files

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.h:
```diff
@@ -0,0 +1,74 @@
+    static constexpr TimeFormatLocalization::CalendarTypeEnum kDefaultCalendarType =
+        TimeFormatLocalization::CalendarTypeEnum::kBuddhist;
+    static constexpr TimeFormatLocalization::HourFormatEnum kDefaultHourFormat = TimeFormatLocalization::HourFormatEnum::k12hr;
+    static constexpr size_t kMaxExpectedAttributeCount                         = 3;
```
It was used to build the Attribute list before integrating `AttributeListBuilder`, remove.

tersal committed:
Add fixes from PR comments

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,222 @@
+
+        // Using WriteAttribute directly so we can check that the decoded value is in the supported list
+        // before storing it.
+        CHIP_ERROR result = attrProvider.WriteValue({ kRootEndpointId, TimeFormatLocalization::Id, ActiveCalendarType::Id },
```
Updated the code and replaced logic layer. Now using NotifyAttributeChangedIfSuccess.

tersal commented on src/app/persistence/AttributePersistence.h:
```diff
@@ -52,10 +52,19 @@ class AttributePersistence
+        ReturnErrorOnFailure(decoder.Decode(decodedValue));
+        if constexpr (std::is_enum_v<T>)
+        {
+            VerifyOrReturnError(decodedValue != T::kUnknownEnumValue, CHIP_ERROR_INVALID_ARGUMENT);
```
Replaced in PR #40568

tersal committed:
Remove additional logic for return error values

tersal committed:
Change attributelist logic

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-logic.cpp:
```diff
@@ -0,0 +1,222 @@
+    case TimeFormatLocalization::Attributes::HourFormat::Id: {
+        CHIP_ERROR result = attrPersistence.DecodeAndStoreNativeEndianValue(
+            { kRootEndpointId, TimeFormatLocalization::Id, HourFormat::Id }, decoder, mHourFormat);
+        return result == CHIP_NO_ERROR ? Protocols::InteractionModel::Status::Success
```
Removed this logic, returning directly the value from `DecodeAndStoreNativeEndianValue`

tersal committed:
Change default value handling for attributes

tersal committed:
Revert ZAP changes for attributes

tersal committed:
Adjust UnitTest for Cluster only layer implementation

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Regenerate files with ZAP

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,237 @@
+                                                             TimeFormatLocalization::HourFormatEnum defaultHourFormat, 
+                                                             TimeFormatLocalization::CalendarTypeEnum defaultCalendarType) :
+    DefaultServerCluster({ endpointId, TimeFormatLocalization::Id }),
+    mFeatures(features), mHourFormat(defaultHourFormat), mCalendarType(defaultCalendarType)
```
The line lengths seem to be corrected by restyle, not sure if I changing the line after applying restyle changes will trigger a CI error.

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,237 @@
+            return Protocols::InteractionModel::Status::ConstraintError;
+        }
+
+         mCalendarType = newCalendar;
```
Fixed with restyle.

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,237 @@
+                                                     ReadOnlyBufferBuilder<DataModel::AttributeEntry> & builder)
+{
+    AttributeListBuilder listBuilder(builder);
+    
```
I think that this change was handled by restyle.

tersal commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,95 @@
+    TimeFormatLocalization::CalendarTypeEnum defaultCalendarType = TimeFormatLocalization::CalendarTypeEnum::kBuddhist;
+
+    // First the HourFormat
+    if(TimeFormatLocalization::Attributes::HourFormat::Get(endpoint, &defaultHourFormat) != Status::Success) 
```
Corrected by restyle.

tersal commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    // Get the ActiveCalendarType default value if the feature is enabled.
+    if(BitFlags<TimeFormatLocalization::Feature>(rawFeatureMap).Has(TimeFormatLocalization::Feature::kCalendarFormat))
```
Corrected by restyle.

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,237 @@
+    chip::app::OptionalAttributeSet<TimeFormatLocalization::Attributes::ActiveCalendarType::Id,
+                                    TimeFormatLocalization::Attributes::SupportedCalendarTypes::Id> optionalAttributeSet;
+
+    if(mFeatures.Has(TimeFormatLocalization::Feature::kCalendarFormat)) 
```
Corrected by restyle.

tersal commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,95 @@
+    // Get the ActiveCalendarType default value if the feature is enabled.
+    if(BitFlags<TimeFormatLocalization::Feature>(rawFeatureMap).Has(TimeFormatLocalization::Feature::kCalendarFormat))
+    {
+        if(TimeFormatLocalization::Attributes::ActiveCalendarType::Get(endpoint, &defaultCalendarType) != Status::Success) 
```
Corrected by restyle.

tersal commented on src/app/persistence/AttributePersistence.h:
```diff
@@ -52,10 +52,19 @@ class AttributePersistence
-    template <typename T, typename std::enable_if_t<std::is_arithmetic_v<T>> * = nullptr>
+    template <typename T, typename std::enable_if_t<std::is_arithmetic_v<T> || std::is_enum_v<T>> * = nullptr>
     CHIP_ERROR DecodeAndStoreNativeEndianValue(const ConcreteAttributePath & path, AttributeValueDecoder & decoder, T & value)
     {
-        ReturnErrorOnFailure(decoder.Decode(value));
+        T decodedValue;
+
+        ReturnErrorOnFailure(decoder.Decode(decodedValue));
+        if constexpr (std::is_enum_v<T>)
+        {
+            VerifyOrReturnError(decodedValue != T::kUnknownEnumValue, CHIP_ERROR_INVALID_ARGUMENT);
+        }
+
+        value = decodedValue;
+
         return mProvider.WriteValue(path, { reinterpret_cast<const uint8_t *>(&value), sizeof(value) });
```
Will remove this code since it will be handled in a different PR.

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    // Delete copy constuctor and assignement
```
Fixed.

tersal committed:
Add review fixes

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,238 @@
+    {
+        // Update the optional validCalendar to a value from the SupportedList.
+        // This will return either the last element of the list or the requested
+        // value if exists.
```
Updated comment.

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Revert changes to AttributePersistence

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

bzbarsky-apple commented on src/app/zap-templates/zcl/zcl-with-test-extensions.json:
```diff
@@ -817,6 +817,7 @@
             "CurrentThermostatSuggestion",
             "ThermostatSuggestionNotFollowingReason"
         ],
+        "Time Format Localization": ["ClusterRevision"],
```
I guess this does not list the other non-list attributes because we use the ZAP RAM store for the featuremap and whatnot?

tersal replied:
Yes, I wasn't able to remove more elements since we still need the values from ZAP, even for the two attributes where we need them for the "default" value just in case.

tersal commented on src/app/clusters/time-format-localization-server/CodegenIntegration.cpp:
```diff
@@ -0,0 +1,96 @@
+    TimeFormatLocalization::CalendarTypeEnum defaultCalendarType = TimeFormatLocalization::CalendarTypeEnum::kBuddhist;
+
+    // First the HourFormat
+    if (TimeFormatLocalization::Attributes::HourFormat::Get(endpoint, &defaultHourFormat) != Status::Success)
```
I think that this is the intention with the issue #40595 , not sure if worded correctly there but the goal should be similar.

andy31415 replied:
My expectation in general would be that we will end up with a "if you want a small app, you will code it as purely code-driven" which will drop all the overhead from CodegenIntegration (which is painful because it reads generated defaults).

CodegenIntegration functions are often quite large and some embedded platforms may not care (e.g. they never do shutdown).

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,238 @@
+
+        mCalendarType = newCalendar;
+
+        // Using WriteAttribute directly so we can check that the decoded value is in the supported list
```
I think it makes sense, currently `DecodeAndStoreNativeEndianValue` only checks for Unknown enum values but there may be cases (like in this cluster) where a more specific check is needed.

I created the issue #40708 as a follow up.

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
minor changes from review

restyled-commits committed:
Restyled by whitespace

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,238 @@
+        TimeFormatLocalization::Attributes::SupportedCalendarTypes::kMetadataEntry
+    };
+
+    chip::app::OptionalAttributeSet<TimeFormatLocalization::Attributes::ActiveCalendarType::Id,
```
Removed.

tersal committed:
Regenerate ZAP files

tersal commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,238 @@
+
+        mCalendarType = newCalendar;
+
+        // Using WriteAttribute directly so we can check that the decoded value is in the supported list
```
Changed to `WriteValue`, since we are using that instead of `DecodeAndStoreNativeEndianValue` due to the additional check needed for this attribute.

tersal committed:
Add validation for empty Calendar list

tersal committed:
Change error code for ReadAttribute

tersal committed:
Update copyright dates for files that use partially the old implementation

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

tersal committed:
Regenerate ZAP files

tersal committed:
Merge branch 'master' into decouple-time-format-cluster

andy31415 commented on src/app/clusters/time-format-localization-server/app_config_dependent_sources.cmake:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2025 Project CHIP Authors
+# Copyright (c) 2021-2025 Project CHIP Authors
```
I think this was not created in the past - the codegen sources was from this year.

andy31415 commented on src/app/clusters/time-format-localization-server/app_config_dependent_sources.gni:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2025 Project CHIP Authors
+# Copyright (c) 2021-2025 Project CHIP Authors
```
```suggestion
# Copyright (c) 2025 Project CHIP Authors
```

This file did not exist in 2021 and it was not copied from a file from 2021 either

andy31415 committed:
Merge with master

andreilitvin committed:
Merge branch 'master' into decouple-time-format-cluster

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,251 @@
+        // it means that the SupportedCalendar list is empty;
+        if (validCalendar == TimeFormatLocalization::CalendarTypeEnum::kUnknownEnumValue)
+        {
+            return CHIP_ERROR_NOT_FOUND;
```
This seems unrecoverable ... maybe we should just log instead?

andy31415 commented on src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp:
```diff
@@ -0,0 +1,238 @@
+    // We could have an invalid calendar type value if an OTA update removed support for the value we were using.
+    // If initial value is not one of the allowed values, pick one valid value and write it.
+    TimeFormatLocalization::CalendarTypeEnum validCalendar = defaultCalendarType;
+    if (!IsSupportedCalendarType(mCalendarType, &validCalendar))
```
I think the constructor provides us a default value. We should use that instead and not try to guess a valid value.

andy31415 committed:
Slight updates for smaller code (30 bytes saved it seems)

andy31415 committed:
Update for slightly smaller code, fix some logic and copyright.

andy31415 commented:
After a set of code updades, this became slightly smaller:

|Type|Size|Function|Size1|Size2|
|-|-|-|-|-|
|REMOVED|-292|(anonymous namespace)::TimeFormatLocalizationAttrAccess::Read|0|292|
|REMOVED|-116|MatterTimeFormatLocalizationClusterServerPreAttributeChangedCallback|0|116|
|REMOVED|-92|(anonymous namespace)::IsSupportedCalendarType|0|92|
|REMOVED|-36|vtable for (anonymous namespace)::TimeFormatLocalizationAttrAccess|0|36|
|REMOVED|-34|_GLOBAL__sub_I__Z68MatterTimeFormatLocalizationClusterServerPreAttributeChangedCallbackRKN4chip3app21ConcreteAttributePathEhtPh|0|34|
|REMOVED|-32|chip::app::Clusters::TimeFormatLocalization::Attributes::ActiveCalendarType::Set|0|32|
|CHANGED|-22|MatterTimeFormatLocalizationPluginServerInitCallback|2|24|
|REMOVED|-12|(anonymous namespace)::AutoReleaseIterator::~AutoReleaseIterator|0|12|
|REMOVED|-10|(anonymous namespace)::AutoReleaseIterator::Next(chip::app::Clusters::TimeFormatLocalization::CalendarTypeEnum&) [clone .isra.0]|0|10|
|REMOVED|-6|(anonymous namespace)::TimeFormatLocalizationAttrAccess::~TimeFormatLocalizationAttrAccess|0|6|
|CHANGED|-2|attributeData|4348|4350|
|ADDED|6|chip::app::Clusters::TimeFormatLocalizationCluster::~TimeFormatLocalizationCluster|6|0|
|ADDED|8|chip::app::Clusters::(anonymous namespace)::kMandatoryAttributes|8|0|
|ADDED|10|chip::app::Clusters::(anonymous namespace)::AutoReleaseIterator::Next(chip::app::Clusters::TimeFormatLocalization::CalendarTypeEnum&) [clone .isra.0]|10|0|
|ADDED|12|chip::app::Clusters::(anonymous namespace)::AutoReleaseIterator::~AutoReleaseIterator|12|0|
|ADDED|36|chip::app::Clusters::TimeFormatLocalizationCluster::WriteAttribute|36|0|
|CHANGED|40|emberAfTimeFormatLocalizationClusterServerInitCallback|136|96|
|ADDED|42|chip::app::Clusters::TimeFormatLocalization::Attributes::HourFormat::Get|42|0|
|ADDED|50|MatterTimeFormatLocalizationClusterServerShutdownCallback|50|0|
|ADDED|62|chip::app::LazyRegisteredServerCluster<chip::app::Clusters::TimeFormatLocalizationCluster>::Destroy|62|0|
|ADDED|66|chip::app::Clusters::TimeFormatLocalizationCluster::Attributes|66|0|
|ADDED|68|vtable for chip::app::Clusters::TimeFormatLocalizationCluster|68|0|
|ADDED|72|chip::app::Clusters::TimeFormatLocalizationCluster::TimeFormatLocalizationCluster|72|0|
|ADDED|78|chip::app::Clusters::(anonymous namespace)::IsSupportedCalendarType|78|0|
|ADDED|158|chip::app::Clusters::TimeFormatLocalizationCluster::Startup|158|0|
|ADDED|234|chip::app::Clusters::TimeFormatLocalizationCluster::WriteImpl|234|0|
|ADDED|342|chip::app::Clusters::TimeFormatLocalizationCluster::ReadAttribute|342|0|
|TOTAL|630||1345440550|1345439920|

andy31415 committed:
Fix load/store logic a bit more.

andy31415 committed:
Restyle

andy31415 committed:
Update src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp

andy31415 committed:
Fix logic to select a supported calendar.

ZAP files seem to use defaults (i.e. 0 i.e. budhist calendar)
which is wrong, but we must handle.

andy31415 committed:
Restyle

andy31415 committed:
Merge branch 'master' into decouple-time-format-cluster

andy31415 committed:
Fix new callbacks for code driven clusters.

andy31415 committed:
Undo the bad merge of shutdown functions.
Sort the code driven cluster list.

andy31415 committed:
Standardize the startup/shutdown for time format localization.

This saves a bit of flash.

andy31415 committed:
ZAP Regen and minor style update.

andy31415 committed:
Update src/app/clusters/time-format-localization-server/time-format-localization-cluster.cpp

andy31415 committed:
Merge branch 'master' into decouple-time-format-cluster

jadhavrohit924 commented on examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:
```diff
@@ -7716,14 +7716,14 @@ endpoint 0 {
-    persist  attribute hourFormat default = 0;
-    persist  attribute activeCalendarType default = 0;
+    ram      attribute hourFormat default = 0;
+    ram      attribute activeCalendarType default = 0;
```
> The reason we have it RAM (rather than callback) is to allow ZAP UI values to pass in the defaults to the application.

We can pass the default value to NVM as well in the ZAP UI.
